pandas/tests/arrays/test_boolean.py

```diff
@@ -0,0 +1,277 @@
+    values = np.array([True, False, True, False], dtype="bool")
+    mask = np.array([False, False, False, True], dtype="bool")
+
+    result = BooleanArray(values, mask)
+    assert result._data is values
+    assert result._mask is mask
+
+    result = BooleanArray(values, mask, copy=True)
+    assert result._data is not values
+    assert result._mask is not mask
+
+
+@pytest.mark.parametrize(
+    "a, b",
+    [
+        ([True, None], [True, np.nan]),
+        ([None], [np.nan]),
+        ([None, np.nan], [np.nan, np.nan]),
+        ([np.nan, np.nan], [np.nan, np.nan]),
+    ],
+)
+def test_to_boolean_array_none_is_nan(a, b):
+    result = pd.array(a, dtype="boolean")
+    expected = pd.array(b, dtype="boolean")
+    tm.assert_extension_array_equal(result, expected)
+
+
+# @pytest.mark.parametrize(
+#     "values",
+#     [
+#         ["foo", "bar"],
+#         ["1", "2"],
+#         "foo",
+#         [1],
+#         [1.0],
+#         pd.date_range("20130101", periods=2),
+#         np.array(["foo"]),
+#         [[1, 2], [3, 4]],
+#         [np.nan, {"a": 1}],
+#     ],
+# )
+# def test_to_boolean_array_error(values):
+#     # error in converting existing arrays to BooleanArray
+#     with pytest.raises(TypeError):
+#         pd.array(values, dtype="boolean")
+
+
+def test_to_boolean_array_integer():
+    result = pd.array([1, 0, 1, 0], dtype="boolean")
+    expected = pd.array([True, False, True, False], dtype="boolean")
+    tm.assert_extension_array_equal(result, expected)
+
+    # with pytest.raises(TypeError, match="cannot safely cast non-equivalent"):
+    #     pd.array([1, 2, 3], dtype="boolean")
+
+
+def test_coerce_to_numpy_array():
+    # with missing values -> object dtype
+    arr = pd.array([True, False, None], dtype="boolean")
+    result = np.array(arr)
+    expected = np.array([True, False, None], dtype="object")
+    tm.assert_numpy_array_equal(result, expected)
+
+    # with no missing values -> bool dtype
+    # TODO consistently return object dtype instead?
+    arr = pd.array([True, False, True], dtype="boolean")
+    result = np.array(arr)
+    expected = np.array([True, False, True], dtype="bool")
+    tm.assert_numpy_array_equal(result, expected)
+
+
+def test_astype():
+    # with missing values
+    arr = pd.array([True, False, None], dtype="boolean")
+    msg = "cannot convert float NaN to"
+
+    with pytest.raises(ValueError, match=msg):
+        arr.astype("int64")
+
+    with pytest.raises(ValueError, match=msg):
+        arr.astype("bool")
+
+    # no missing values
+    arr = pd.array([True, False, True], dtype="boolean")
+    # result = arr.astype("int64")
+    # expected = np.array([1, 0, 1], dtype="int64")
+    # tm.assert_numpy_array_equal(result, expected)
+
+    result = arr.astype("bool")
+    expected = np.array([True, False, True], dtype="bool")
+    tm.assert_numpy_array_equal(result, expected)
+
+
+# @pytest.mark.parametrize("ufunc", [np.add])
+# def test_ufuncs_binary(ufunc):
```

Comment:
still need to implement arithmetic ops

Reply:
What are the expected return types? Would you use IntegerArray for things returning an integer dtype?

ndarray[bool] will sometimes return a float dtype...

```python
In [17]: a = np.array([True, False])

In [18]: a / 1
Out[18]: array([1., 0.])
```

would we return an ndarray there, with NaN for the missing values?

Reply:
I guess returning a PandasArray is also an option... If we want to stay in the algebra of `op(BooleanArray, other)` returning an ExtensionArray

Reply:
For now I went with tthe dtype that numpy ops result in, and then if result is bool -> BooleanArray, if is int -> IntegerArray, otherwise if result is float dtype (or something else) return the numpy array.